### PR TITLE
207 click on graph legends to perform actions

### PR DESCRIPTION
**Pull request by vijulshah**
Single-click on legends to compare experiments on the same page.
Instead of a double-click feature, gave a drop-down to select the experiment id and navigate to the experiment page.

### CODE REVIEW

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx [hunk 1]
```diff
@@ -43,8 +46,58 @@ export default function Graph({ chart_id, exp_id, exp_data }: { chart_id: string
             })),
     };
 
+    const legendOnClickHandler = function (event:any, legendItem:any, legend:any) {
+        let index = legendItem.datasetIndex;
+        let ci = legend.chart;
+        let meta = ci.getDatasetMeta(index);
+
+        // Get the indexes of datasets with hidden === false
+        let visibleIndexes = ci.data.datasets.reduce(function(acc:any, dataset:any, i:any) {
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:55]: making a variable a **const** is not just better for developer not to overwrite it later but also better for the reviewer, I know then that this variable is always going to point to the same object or instance in memory generally speaking!

and that's everywhere where not just here 🥲

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx [hunk 2]
```diff
@@ -14,8 +17,8 @@ export default function Graph({ chart_id, exp_id, exp_data }: { chart_id: string
 
     const chartLabels = useAppSelector(state => selectChartLabelsById(state, chart_id));
     const experimentsDict = useAppSelector(state => selectExperimentsPerChartById(state, chart_id));
-
-    // prepare data (Warning Looping!)
+    const location = useLocation();
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:20]: is **useLocation** better or using the **idTab** in the **GlobalConfig** redux slice better? and isn't one of them is redundant then?

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx [hunk 3]
```diff
@@ -43,8 +46,58 @@ export default function Graph({ chart_id, exp_id, exp_data }: { chart_id: string
...
+        // Toggle visibility of clicked dataset
+        if(meta.hidden === false) {
+            if(visibleIndexes.length === 1) {
+                ci.data.datasets.forEach(function(dataset:any, i:any) {
+                    let otherMeta = ci.getDatasetMeta(i);
+                    otherMeta.hidden = null;
+                });
+            }
+            else {
+                if(meta.hidden === false) {
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:72]: this if condition is redundant, no?

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx [hunk 4]
```diff
@@ -43,8 +46,58 @@ export default function Graph({ chart_id, exp_id, exp_data }: { chart_id: string
...
+        }
+        else {
+            meta.hidden = false;
+            ci.data.datasets.forEach(function(dataset:any, i:any) {
+                let otherMeta = ci.getDatasetMeta(i);
+                if (otherMeta.hidden === null && i !== index) {
+                    otherMeta.hidden = true;
+                }
+            });
+        }
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:85]: adding maybe a bit  more comments here to explain how exactly it "Toggle visibility of clicked dataset" ?

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx [hunk 5]
```diff
@@ -98,6 +169,12 @@ export default function Graph({ chart_id, exp_id, exp_data }: { chart_id: string
                         options={options}
                     />
                 </Box>
+                {
+                    location.pathname.split("/")[1] !== RoutesMapping["ExperimentPage"].url ?
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:173]: **location** vs **redux** vs **localstorage** ?

File: src/ml_board/frontend/Dashboard/src/components/graphs/SelectExperimentDropdown/SelectExperimentDropdown.tsx
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/graphs/SelectExperimentDropdown/SelectExperimentDropdown.tsx:1]: good work here and also for making a separate component 👍

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx [hunk 6]
```diff
@@ -42,19 +46,21 @@ const Settings: React.FC<FuncProps> = (props) => {
                 setConfigTextState(settingConfigs);
             });
         }
-    },[])
+    },[isSocketConnected && (grid_search_id || rest_api_url || socket_connection_url)])
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:49]: `[isSocketConnected , grid_search_id , rest_api_url , socket_connection_url]` ?

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx [hunk 7]
```diff
@@ -15,14 +15,18 @@ export const defaultSocketConnectionUrlHelperText = "eg: http://127.0.0.1:5002";
 export const defaultRestApiUrlHelperText = "eg: http://127.0.0.1:5001";
 
 export interface FuncProps {
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:17]: are we using **FuncProps** elsewhere other than this file? if not then we shouldn't export it

File: src/ml_board/frontend/Dashboard/src/components/settings/SelectExperimentDropdown/SelectExperimentDropdown.tsx
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/settings/SelectExperimentDropdown/SelectExperimentDropdown.tsx:1]: I was wondering why was this file deleted 👀 was it renamed?

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx [hunk 8]
```diff
@@ -15,14 +15,18 @@ export const defaultSocketConnectionUrlHelperText = "eg: http://127.0.0.1:5002";
...
+    isConfigValidated?: boolean
 }
 
 // Settings file and Config popup file (except render part) might look exactly the same for now. But Settigns will have more functionality later. So we will add more things in Settings which will make it different from popup file.
 const Settings: React.FC<FuncProps> = (props) => {
 
     const isSocketConnected = useAppSelector(isConnected);
+    const grid_search_id = useAppSelector(getGridSearchId);
+    const rest_api_url = useAppSelector(getRestApiUrl);
+    const socket_connection_url = useAppSelector(getSocketConnectionUrl);
```
Comment (OsamaMSoliman) [inline src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:29]: something about the circulation of these 3 doesn't look good but they are working for now, so I accept them now but maybe we will change them in the future!

**Review by OsamaMSoliman — APPROVED** [review]

**Review by OsamaMSoliman — CHANGES_REQUESTED** [review]
@vijulshah small changes are required, I marked them with `[X]`, otherwise can be done later

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx (hunk 3 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:72]: yes, it's redundant. Removed it!

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx (hunk 5 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:173]: removed location from everywhere as said above. Local storage is used for auto-populating the config popup as we discussed in the meeting.

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx (hunk 8 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:29]: Okay!

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx (hunk 7 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:17]: Yes, we r using it in ConfigPopup.tsx

File: src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx (hunk 4 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/components/graphs/Graph.tsx:85]: Added!

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx (hunk 6 above)
Comment (vijulshah) [reply src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx:49]: instead of the "OR ( || )" condition, it's a good practice to write like this. And while making changes I figured we didn't need "&" condition

**Comment by vijulshah** [comment]
@OsamaMSoliman  I have resolved conflicts & made changes as per your review. Pls have a look and merge it if everything looks good

File: src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx (hunk 6 above)
Comment (OsamaMSoliman) [reply src/ml_board/frontend/Dashboard/src/components/settings/Settings.tsx]: This part disappeared, I can see it in the **Files changed** section of the PR ?!

**Review by OsamaMSoliman — APPROVED** [review]
I ran the code, all looks good, this closes #207
